test/studies/shootout/submitted/knucleotide4.chpl

```diff
@@ -0,0 +1,140 @@
+/* The Computer Language Benchmarks Game
+   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
+
+   contributed by Ben McDonald
+   derived from the Chapel #3 version by Ben Harshbarger and Brad Chamberlain
+*/
+
+use IO, Map, Sort;
+
+config param columns = 61;
+
+
+proc main(args: [] string) {
+  // Open stdin and a binary reader channel
+  const consoleIn = openfd(0),
+        fileLen = consoleIn.size,
+        stdinNoLock = consoleIn.reader(kind=ionative, locking=false);
+
+  // Read line-by-line until we see a line beginning with '>TH'
+  var buff: [1..columns] uint(8),
+      lineSize = 0,
+      numRead = 0;
+
+  while stdinNoLock.readline(buff, lineSize) && !startsWithThree(buff) do
+    numRead += lineSize;
+
+  // Read in the rest of the file
+  var dataDom = {1..fileLen-numRead},
+      data: [dataDom] uint(8),
+      idx = 1;
+
+  while stdinNoLock.readline(data, lineSize, idx) do
+    idx += lineSize - 1;
+
+  // Resize our array to the amount actually read
+  dataDom = {1..idx};
+
+  // Make everything uppercase
+  forall d in data do
+    d -= ("a".toByte() - "A".toByte());
+
+  writeFreqs(data, 1);
+  writeFreqs(data, 2);
+  writeCount(data, "GGT");
+  writeCount(data, "GGTA");
+  writeCount(data, "GGTATT");
+  writeCount(data, "GGTATTTTAATT");
+  writeCount(data, "GGTATTTTAATTTATAGT");
+}
+
+
+proc writeFreqs(data, param nclSize) {
+  const freqs = calculate(data, nclSize);
+
+  // create an array of (frequency, sequence) tuples
+  var arr = for (s,f) in freqs.items() do (f,s);
+
+  // print the array, sorted by decreasing frequency
+  for (f, s) in arr.sorted(reverseComparator) do
+   writef("%s %.3dr\n", decode(s, nclSize),
+           (100.0 * f) / (data.size - nclSize));
+  writeln();
+}
+
+
+proc writeCount(data, param str) {
+  const strBytes = str.bytes(),
+        freqs = calculate(data, str.numBytes),
+        d = hash(strBytes, strBytes.domain.low, str.numBytes);
+
+  writeln(freqs[d], "\t", decode(d, str.numBytes));
+}
+
+
+proc calculate(data, param nclSize) {
+  var freqs = new map(hashVal, int);
+
+  var lock: sync bool = true;
+  const numTasks = here.maxTaskPar;
+  coforall tid in 1..numTasks with (ref freqs) {
+    var myFreqs = new map(hashVal, int);
+
+    for i in tid..(data.size-nclSize) by numTasks do
+      myFreqs[hash(data, i, nclSize)] += 1;
+
+    lock.readFE();      // acquire lock
+    for (k,v) in myFreqs.items() do
+      freqs[k] += v;
+    lock.writeEF(true); // release lock
+  }
+
+  return freqs;
+}
+
+
+const toChar: [0..3] string = ["A", "C", "T", "G"];
+var toNum: [0..127] int;
+
+forall i in toChar.domain do
+  toNum[toChar[i].toByte()] = i;
+
+
+inline proc decode(in data, param nclSize) {
+  var ret: string;
+
+  for i in 1..nclSize {
+    ret = toChar[(data.val & 3)] + ret;
+    data.val >>= 2;
```

Comment:
This PR generally looks good to me, where the only comment I can think of would be to have `decode()` still take in an `int` so that it is exactly the same as before and any unwrapping of `hashVal`s is done at the callsites.  Specifically... (see next comment)